<!DOCTYPE html>
<html>
  <head>
    <meta name="title" content="About">
    <meta name="date" content="2015-12-05">
    <meta name="slug" content="about">
  </head>
  <body>
    <div class="container">
      <div class="col-md-8 page-header">
        <h1>About</h1>
      </div>
      <div class="row">
        <div class="col-md-8">
          <p>

            I'm Vince Buffalo, a Ph.D. student in <a
href="http://gcbias.org/">Graham Coop</a>'s lab at UC Davis in the <a
href="http://www.cpb.ucdavis.edu/">Center for Population Biology</a>, studying
evolutionary and population genetics. Previously, I worked as a
bioinformatician at the <a href="http://genomecenter.ucdavis.edu/">UC Davis
Genome Center</a> and in the <a
href="http://dubcovskylab.ucdavis.edu">Dubcovsky</a> and <a
href="http://www.rilab.org/">Ross-Ibarra</a> labs. I wrote <a href="/book/">Bioinformatics Data Skills</a> published by O'Reilly Media in July
2015. I use statistics, probability, and computers to learn about evolution.

          </p>

          <h2>The blind play of genes</h2>

          <p>

            I named my site "the blind play of genes" after a sentence in G.C.
            Williams' <em>Adaptation and Natural Selection</em>:

            <blockquote class="tiny-blockquote"> 

              The opposition [to natural selection] arises, as Darwin himself
              observed, not from what reason dictates but from the limits of
              what the imagination can accept. It is difficult for many people
              to imagine that an individual's role in evolution is entirely
              contained in its contribution to vital statistics. It is
              difficult to imagine that an acceptable moral order could arise
              from vital statistics, and difficult to dispense with belief in a
              moral order in living nature.  <strong>It is difficult to imagine
                that the blind play of genes could produce man.</strong>
              <footer style="text-align: right;">G.C. Williams, <em>Adaptation
                  and natural Selection</em>, p. 3-4</footer>
            </blockquote>

          </p>
        </div>
        <div class="col-md-3" style="margin-left: 3em;">
          <img alt="Photograph of Vince Buffalo" src="/images/me.png" width="230" class="media-object" />
        </div>


      </div><!-- row -->
      <h2>Curriculum Vitae Summary</h2>
      <!-- CV -->
      <div class="row">
        <div class="col-md-4 cv-left-col">
          Education
        </div>
        <div class="col-md-7">
          <p>Ph.D. in Population Genetics, UC Davis, <em>expected 2018</em></p>
          <p>B.A. in Economics and Political Science, UC Davis, 2009<br/>
            <span style="font-size: 0.8em;">Minor in Statistics<br/>
              Phi Beta Kappa, Dean’s List, Departmental Citation in Economics<br/>
            </span>
          </p>
        </div>
      </div>


      <div class="row">
        <div class="col-md-4 cv-left-col">
          Book
        </div>
        <div class="col-md-7">
          <p><em>Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools</em>, O’Reilly Media, July 2015</p>
        </div>
      </div>

      <div class="row">
        <div class="col-md-4 cv-left-col">
          Fellowships and Awards
        </div>
        <div class="col-md-7">
          <p>
            2014 - NSF Graduate Research Fellowship Program<br/>
            2013 - NSF Graduate Research Fellowship Program, Honorable Mention<br/>
            2013 - UC Davis Graduate Scholars Fellowship<br/>
          </p>
        </div>
      </div>

      <div class="row">
        <div class="col-md-4 cv-left-col">
          Professional Research Experience
        </div>
        <div class="col-md-7">

          <strong>Staff Computational Biologist, Jeffrey Ross-Ibarra
            Lab</strong>  ─ 9/2013 to 9/2014<br/>

          <span style="font-size: 0.8em;"> 
            <p>
              UC Davis Department of Plant Sciences Research projects:
              investigating genetic load in wild outcrossing teosinte populations
              with next-generation sequencing and GBS; development of imputation,
              phasing, and identification of rare alleles statistical genetics
              methods; short-read aligner asssesment in highly paralogous
              genomes.
            </p>
          </span>
          <strong>Bioinformatician, Jorge Dubcovsky Lab</strong> ─ 7/2013 to 9/2013<br/>
          <span style="font-size: 0.8em;"> 
            <p>

              Research projects: developing novel software to extend
              single-individual haplotyping methods to phase homeologs in
              tetraploid wheat transcriptome data, development of a robust
              and sensitive RNA-seq bioinformatics protocol in tetraploid
              wheat.

            </p>
          </span>
          <strong>Lead Statistical Programmer, Bioinformatics Core</strong> ─ 9/2009 to 7/2013<br/>
          <span style="font-size: 0.8em;"> 
            <p>

              Research projects: developing next-generation sequencing quality
              pipeline, applying machine learning approaches (elastic net,
              lasso, sparse PCA) to RNA-seq expression data, RNA-seq
              statistical analysis, developed novel software to identify
              chromosomal translocations in post-chemotherapy patients using
              next-generation data.
          </p> </span>
        </div>
      </div>


    </div>




  </div><!-- container -->
</body>
</html>
